From a156909af1d448baca4e564f8216e985a5a7fc1f Mon Sep 17 00:00:00 2001
From: Stanley Chu <yschu@nuvoton.com>
Date: Tue, 4 Nov 2025 10:00:49 +0800
Subject: [PATCH] add jtm spi driver

DTS example:
&jtm1 {
        status = "okay";
        compatible = "nuvoton,npcm845-jtm-spi";
        #address-cells = <1>;
        #size-cells = <0>;
        cs-gpios = <&gpio0 0 GPIO_ACTIVE_LOW>; /* gpio0 */
	//use-polling; // enable polling mode
        Flash@0 {
                compatible = "jedec,spi-nor";
                reg = <0x0>;
                #address-cells = <1>;
                #size-cells = <1>;
                spi-max-frequency = <1000000>;
                partition@0 {
                     label = "jtm_spi_flash";
                        reg = <0x0 0x0>;
                };
       };

};

Upstream-Status: Pending [Not submitted to upstream yet]
Signed-off-by: Stanley Chu <yschu@nuvoton.com>
---
 drivers/spi/Kconfig        |   7 +
 drivers/spi/Makefile       |   1 +
 drivers/spi/spi-npcm-jtm.c | 460 +++++++++++++++++++++++++++++++++++++
 3 files changed, 468 insertions(+)
 create mode 100644 drivers/spi/spi-npcm-jtm.c

diff --git a/drivers/spi/Kconfig b/drivers/spi/Kconfig
index 0b4317ebfd7b..e7cdcb5a41e0 100644
--- a/drivers/spi/Kconfig
+++ b/drivers/spi/Kconfig
@@ -752,6 +752,13 @@ config SPI_NPCM_FLASH
 	help
 	  Enable Nuvoton ESPI Flash
 
+config SPI_NPCM_JTM
+	tristate "Nuvoton NPCM JTM SPI Controller"
+	depends on ARCH_NPCM || COMPILE_TEST
+	help
+	  This driver provides support for Nuvoton NPCM BMC
+	  JTM SPI controller in master mode.
+
 config SPI_LANTIQ_SSC
 	tristate "Lantiq SSC SPI controller"
 	depends on LANTIQ || X86 || COMPILE_TEST
diff --git a/drivers/spi/Makefile b/drivers/spi/Makefile
index d20ae3846d10..91399c531cce 100644
--- a/drivers/spi/Makefile
+++ b/drivers/spi/Makefile
@@ -94,6 +94,7 @@ obj-$(CONFIG_SPI_MXIC)			+= spi-mxic.o
 obj-$(CONFIG_SPI_MXS)			+= spi-mxs.o
 obj-$(CONFIG_SPI_WPCM_FIU)		+= spi-wpcm-fiu.o
 obj-$(CONFIG_SPI_NPCM_FIU)		+= spi-npcm-fiu.o
+obj-$(CONFIG_SPI_NPCM_JTM)		+= spi-npcm-jtm.o
 obj-$(CONFIG_SPI_NPCM_PSPI)		+= spi-npcm-pspi.o
 obj-$(CONFIG_SPI_NPCM_FLASH)		+= spi-npcm-flash.o
 obj-$(CONFIG_SPI_NXP_FLEXSPI)		+= spi-nxp-fspi.o
diff --git a/drivers/spi/spi-npcm-jtm.c b/drivers/spi/spi-npcm-jtm.c
new file mode 100644
index 000000000000..a0f0b0598710
--- /dev/null
+++ b/drivers/spi/spi-npcm-jtm.c
@@ -0,0 +1,460 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2024 Nuvoton Technology corporation.
+
+#include <linux/kernel.h>
+#include <linux/bitfield.h>
+#include <linux/bitops.h>
+#include <linux/clk.h>
+#include <linux/interrupt.h>
+#include <linux/io.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/platform_device.h>
+#include <linux/reset.h>
+#include <linux/regmap.h>
+#include <linux/mfd/syscon.h>
+#include <linux/spi/spi.h>
+
+struct npcm_jtm_spi {
+	struct spi_controller *master;
+	struct device *dev;
+	struct reset_control *reset;
+	struct completion xfer_done;
+	struct clk *clk;
+	void __iomem *base;
+	spinlock_t lock;
+	const char *tx_buf;
+	char *rx_buf;
+	u32 tx_bytes;
+	u32 rx_bytes;
+	u32 byte_cnt;
+	u32 speed_hz;
+	bool use_polling;
+};
+
+#define DRIVER_NAME "npcm-jtm_spi"
+
+/* JTM registers */
+#define JTM_CTL			0x00
+#define JTM_STAT		0x04
+#define JTM_CMD			0x08
+#define JTM_TDO_OUT(n)	(0x10 + n * 4)
+#define JTM_TDO_EN(n)	(0x20 + n * 4)
+#define JTM_TDI_IN(n)	(0x40 + n * 4)
+
+#define JTM_CTL_JTM_EN	BIT(0)
+#define JTM_CTL_DONE_IE	BIT(4)
+#define JTM_CTL_TRST	BIT(8)
+#define JTM_CTL_CKDV	GENMASK(23, 16)
+#define JTM_STAT_BUSY	BIT(0)
+#define JTM_STAT_DONE	BIT(1)
+#define JTM_CMD_ST_OP	BIT(0)
+#define JTM_CMD_CK_CNT	GENMASK(15, 8)
+
+#define NPCM_JTM_MAX_RATE		(50000000)
+#define NPCM_JTM_MAX_CKDIV		255
+#define NPCM_JTM_FIFO_SIZE		16	/* bytes */
+#define NPCM_JTM_TIMEOUT_MS		60000
+#define NPCM_PSPI_DEFAULT_CLK		20000000
+
+static unsigned char reverse[16] = {
+	0x0, 0x8, 0x4, 0xC, 0x2, 0xA, 0x6, 0xE,
+	0x1, 0x9, 0x5, 0xD, 0x3, 0xB, 0x7, 0xF
+};
+
+#define REVERSE(x)  ((reverse[((x) & 0x0f)] << 4) | reverse[((x) & 0xf0) >> 4])
+
+static u32 reverse32(u32 input)
+{
+	u32 output;
+	u8 *out = (u8 *)&output;
+	u8 *in = (u8 *)&input;
+	int i;
+
+	for (i = 0; i < 4; i++)
+		out[i] = REVERSE(in[i]);
+
+	return output;
+}
+
+static void npcm_jtm_spi_set_baudrate(struct npcm_jtm_spi *priv, unsigned int speed)
+{
+	u32 ckdiv;
+	u32 regtemp;
+	u32 freq;
+
+	dev_dbg(priv->dev, "%s: speed %u\n", __func__, speed);
+	if (speed > NPCM_JTM_MAX_RATE)
+		return;
+	freq = clk_get_rate(priv->clk);
+	/* ckdiv = DIV_ROUND_CLOSEST(freq, (2 * speed)) - 1; */
+	ckdiv = DIV_ROUND_UP(freq, (2 * speed)) - 1;
+	if (ckdiv > NPCM_JTM_MAX_CKDIV)
+		ckdiv = NPCM_JTM_MAX_CKDIV;
+
+	regtemp = readl(priv->base + JTM_CTL);
+	regtemp &= ~JTM_CTL_CKDV;
+	writel(regtemp | (ckdiv << 16), priv->base + JTM_CTL);
+
+	dev_dbg(priv->dev, "clk freq = %u (clkdiv=%d)\n", (freq / ((ckdiv + 1) * 2)), ckdiv);
+}
+
+static int npcm_jtm_spi_send(struct npcm_jtm_spi *priv)
+{
+	u8 tx_buf[NPCM_JTM_FIFO_SIZE];
+	u32 *tdo32 = (u32 *)&tx_buf[0];
+	int words, bytes;
+	int print_size;
+	u8 *tdo8;
+	u32 cnt;
+	u32 val;
+	int n, i;
+
+	dev_dbg(priv->dev, "%s: bytes %d\n", __func__, priv->tx_bytes);
+	if (priv->tx_bytes > NPCM_JTM_FIFO_SIZE)
+		cnt = NPCM_JTM_FIFO_SIZE;
+	else
+		cnt = priv->tx_bytes;
+
+	words = cnt / 4;
+	bytes = cnt % 4;
+
+	if (priv->tx_buf) {
+		print_size = cnt > 16 ? 16 : cnt;
+		print_hex_dump_debug("jtm O:", DUMP_PREFIX_NONE, 16, 1, priv->tx_buf,
+			     print_size, false);
+		for (i = 0; i < cnt; i++)
+			tx_buf[i] = REVERSE(priv->tx_buf[i]);
+	}
+
+	for (n = 0; n < words; n++) {
+		if (priv->tx_buf) {
+			writel(*tdo32, priv->base + JTM_TDO_OUT(n));
+			tdo32++;
+		} else
+			writel(0, priv->base + JTM_TDO_OUT(n));
+	}
+
+	if (bytes) {
+		tdo8 = (u8 *)tdo32;
+		val = 0;
+		for (i = 0; i < bytes; i++) {
+			if (priv->tx_buf)
+				val |= tdo8[i] << (i * 8);
+		}
+		writel(val, priv->base + JTM_TDO_OUT(n));
+	}
+
+	priv->byte_cnt = cnt;
+	priv->tx_bytes -= cnt;
+	if (priv->tx_buf)
+		priv->tx_buf += cnt;
+
+	/* Start */
+	val = readl(priv->base + JTM_CMD);
+	val &= ~JTM_CMD_CK_CNT;
+	val |= ((cnt * 8) << 8) | JTM_CMD_ST_OP;
+	writel(val, priv->base + JTM_CMD);
+
+	return cnt;
+}
+
+static int npcm_jtm_spi_recv(struct npcm_jtm_spi *priv, u32 cnt)
+{
+	u32 *buf32 = (u32 *)priv->rx_buf;
+	u8 *buf;
+	u32 val;
+	int n, i;
+	int words, bytes;
+
+	if (priv->rx_bytes < cnt)
+		return -EINVAL;
+
+	dev_dbg(priv->dev, "%s: bytes %u\n", __func__, cnt);
+	words = cnt / 4;
+	bytes = cnt % 4;
+	for (n = 0; n < words; n++) {
+		val = readl(priv->base + JTM_TDI_IN(n));
+		if (buf32) {
+			*buf32 = reverse32(val);
+			buf32++;
+		}
+	}
+
+	if (bytes) {
+		buf = (u8 *)buf32;
+		val = readl(priv->base + JTM_TDI_IN(n));
+		if (buf)
+			for (i = 0; i < bytes; i++)
+				buf[i] = REVERSE((val >> (i * 8)) & 0xFF);
+	}
+	priv->rx_bytes -= cnt;
+	if (priv->rx_buf)
+		priv->rx_buf += cnt;
+
+	return 0;
+}
+
+static int npcm_jtm_spi_transfer_all(struct npcm_jtm_spi *priv)
+{
+	u32 reg;
+	int ret;
+
+	while (priv->tx_bytes) {
+		ret = readl_poll_timeout(priv->base + JTM_STAT, reg,
+					((reg & JTM_STAT_BUSY) == 0), 0, 1000000);
+		if (ret) {
+			dev_info(priv->dev, "jtm is busy\n");
+			return ret;
+		}
+		npcm_jtm_spi_send(priv);
+		ret = readl_poll_timeout(priv->base + JTM_STAT, reg,
+					(reg & JTM_STAT_DONE), 0, 1000000);
+		if (ret) {
+			dev_info(priv->dev, "send timeout\n");
+			return ret;
+		}
+		writel(JTM_STAT_DONE, priv->base + JTM_STAT);
+		if (priv->rx_bytes && priv->byte_cnt)
+			npcm_jtm_spi_recv(priv, priv->byte_cnt);
+		cond_resched();
+	}
+
+	return 0;
+}
+
+static irqreturn_t npcm_jtm_spi_handler(int irq, void *dev_id)
+{
+	struct npcm_jtm_spi *priv = dev_id;
+	u32 stat;
+
+	stat = readl(priv->base + JTM_STAT);
+	if (!priv->tx_buf && !priv->rx_buf) {
+		dev_dbg(priv->dev, "no buf\n");
+		writel(stat, priv->base + JTM_STAT);
+		return IRQ_HANDLED;
+	}
+
+	if (priv->tx_bytes == 0 && priv->rx_bytes == 0) {
+		dev_dbg(priv->dev, "no len\n");
+		writel(stat, priv->base + JTM_STAT);
+		complete(&priv->xfer_done);
+		return IRQ_HANDLED;
+	}
+
+	if (stat & JTM_STAT_DONE) {
+		writel(JTM_STAT_DONE, priv->base + JTM_STAT);
+		if (priv->rx_bytes && priv->byte_cnt)
+			npcm_jtm_spi_recv(priv, priv->byte_cnt);
+		if (priv->rx_bytes == 0)
+			complete(&priv->xfer_done);
+	}
+	if (((stat & JTM_STAT_BUSY) == 0)) {
+		if (priv->tx_bytes)
+			npcm_jtm_spi_send(priv);
+	}
+
+	return IRQ_HANDLED;
+}
+
+static int npcm_jtm_spi_transfer_one(struct spi_controller *master,
+				     struct spi_device *spi,
+				     struct spi_transfer *t)
+{
+	struct npcm_jtm_spi *priv = spi_controller_get_devdata(master);
+	//unsigned long flags;
+	int print_size;
+	int ret = 0;
+	u32 val;
+	u8 stat;
+
+	if (!t->len || t->speed_hz > NPCM_JTM_MAX_RATE)
+		return -EINVAL;
+
+	dev_dbg(priv->dev, "%s: len %d\n", __func__, t->len);
+	if (priv->speed_hz != t->speed_hz) {
+		npcm_jtm_spi_set_baudrate(priv, t->speed_hz);
+		priv->speed_hz = t->speed_hz;
+	}
+
+	print_size = t->len > 16 ? 16 : t->len;
+	priv->tx_buf = t->tx_buf;
+	priv->rx_buf = t->rx_buf;
+	priv->tx_bytes = t->len;
+	priv->rx_bytes = t->len;
+
+	stat = readl(priv->base + JTM_STAT);
+	if ((stat & JTM_STAT_BUSY) != 0) {
+		dev_err(priv->dev, "jtm state busy\n");
+		ret = -EBUSY;
+		goto quit;
+	}
+
+	reinit_completion(&priv->xfer_done);
+	/* Enable module and interrupt */
+	val = readl(priv->base + JTM_CTL);
+	val |= JTM_CTL_JTM_EN;
+	if (!priv->use_polling)
+		val |= JTM_CTL_DONE_IE;
+	writel(val, priv->base + JTM_CTL);
+
+	if (priv->use_polling) {
+		ret = npcm_jtm_spi_transfer_all(priv);
+	} else {
+		//spin_lock_irqsave(&priv->lock, flags);
+		npcm_jtm_spi_send(priv);
+		//spin_unlock_irqrestore(&priv->lock, flags);
+
+		ret = wait_for_completion_timeout(&priv->xfer_done,
+					     msecs_to_jiffies
+					     (NPCM_JTM_TIMEOUT_MS));
+		if (ret == 0) {
+			dev_err(priv->dev, "%s: timeout, remaining tx %u rx %u\n",
+				__func__, priv->tx_bytes, priv->rx_bytes);
+			ret = -ETIMEDOUT;
+		} else {
+			ret = 0;
+		}
+	}
+
+	/* Disable module and interrupt */
+	val &= ~(JTM_CTL_JTM_EN | JTM_CTL_DONE_IE);
+	writel(val, priv->base + JTM_CTL);
+quit:
+	if (t->rx_buf) {
+		priv->rx_buf = t->rx_buf;
+		print_hex_dump_debug("jtm I:", DUMP_PREFIX_NONE, 16, 1, priv->rx_buf,
+			     print_size, false);
+	}
+	dev_dbg(priv->dev, "%s: ret %d\n", __func__, ret);
+	return ret;
+}
+
+static void npcm_jtm_spi_reset_hw(struct npcm_jtm_spi *priv)
+{
+	reset_control_assert(priv->reset);
+	udelay(5);
+	reset_control_deassert(priv->reset);
+}
+
+static int npcm_jtm_spi_probe(struct platform_device *pdev)
+{
+	struct npcm_jtm_spi *priv;
+	struct spi_controller *master;
+	unsigned long clk_hz;
+	int irq;
+	int ret;
+
+	master = spi_alloc_master(&pdev->dev, sizeof(*priv));
+	if (!master)
+		return -ENOMEM;
+
+	platform_set_drvdata(pdev, master);
+
+	priv = spi_controller_get_devdata(master);
+	priv->master = master;
+	priv->dev = &pdev->dev;
+
+	priv->base = devm_platform_ioremap_resource(pdev, 0);
+	if (IS_ERR(priv->base)) {
+		ret = PTR_ERR(priv->base);
+		goto out_master_put;
+	}
+
+	priv->clk = devm_clk_get(&pdev->dev, NULL);
+	if (IS_ERR(priv->clk)) {
+		dev_err(&pdev->dev, "failed to get clock\n");
+		ret = PTR_ERR(priv->clk);
+		goto out_master_put;
+	}
+
+	ret = clk_prepare_enable(priv->clk);
+	if (ret)
+		goto out_master_put;
+
+	irq = platform_get_irq(pdev, 0);
+	if (irq < 0) {
+		ret = irq;
+		goto out_disable_clk;
+	}
+
+	priv->reset = devm_reset_control_get(&pdev->dev, NULL);
+	if (IS_ERR(priv->reset)) {
+		ret = PTR_ERR(priv->reset);
+		goto out_disable_clk;
+	}
+
+	if (of_property_read_bool(pdev->dev.of_node, "use-polling")) {
+		priv->use_polling = true;
+		dev_info(priv->dev, "use polling\n");
+	}
+
+	/* reset SPI-HW block */
+	npcm_jtm_spi_reset_hw(priv);
+
+	ret = devm_request_irq(&pdev->dev, irq, npcm_jtm_spi_handler, 0,
+			       "npcm-jtm_spi", priv);
+	if (ret) {
+		dev_err(&pdev->dev, "failed to request IRQ\n");
+		goto out_disable_clk;
+	}
+
+	init_completion(&priv->xfer_done);
+
+	clk_hz = clk_get_rate(priv->clk);
+
+	master->max_speed_hz = 50000000;
+	master->min_speed_hz = 1000000;
+	master->mode_bits = SPI_CPHA;
+	master->dev.of_node = pdev->dev.of_node;
+	master->bus_num = -1;
+	master->bits_per_word_mask = SPI_BPW_MASK(8);
+	master->transfer_one = npcm_jtm_spi_transfer_one;
+	master->use_gpio_descriptors = true;
+
+	/* set to default clock rate */
+	npcm_jtm_spi_set_baudrate(priv, NPCM_PSPI_DEFAULT_CLK);
+
+	ret = devm_spi_register_controller(&pdev->dev, master);
+	if (ret)
+		goto out_disable_clk;
+
+	pr_info("NPCM JTM SPI %d probed\n", master->bus_num);
+
+	return 0;
+
+out_disable_clk:
+	clk_disable_unprepare(priv->clk);
+
+out_master_put:
+	spi_controller_put(master);
+	return ret;
+}
+
+static void npcm_jtm_spi_remove(struct platform_device *pdev)
+{
+	struct spi_controller *master = platform_get_drvdata(pdev);
+	struct npcm_jtm_spi *priv = spi_controller_get_devdata(master);
+
+	npcm_jtm_spi_reset_hw(priv);
+	clk_disable_unprepare(priv->clk);
+}
+
+static const struct of_device_id npcm_jtm_spi_match[] = {
+	{ .compatible = "nuvoton,npcm845-jtm-spi", .data = NULL },
+	{}
+};
+MODULE_DEVICE_TABLE(of, npcm_jtm_spi_match);
+
+static struct platform_driver npcm_jtm_spi_driver = {
+	.driver		= {
+		.name		= DRIVER_NAME,
+		.of_match_table	= npcm_jtm_spi_match,
+	},
+	.probe		= npcm_jtm_spi_probe,
+	.remove		= npcm_jtm_spi_remove,
+};
+module_platform_driver(npcm_jtm_spi_driver);
+
+MODULE_DESCRIPTION("NPCM JTM SPI Controller driver");
+MODULE_LICENSE("GPL v2");
-- 
2.34.1

